File: clean_data.js
const fs = require('fs');

function cleanCsv(filePath, outputPath) {
    const data = fs.readFileSync(filePath, 'utf8');
    const lines = data.split(/\r?\n/);
    const header = lines[0].split(',');
    
    const rows = lines.slice(1).filter(line => line.trim() !== '');
    
    const parseDate = (dateStr) => {
        if (!dateStr || dateStr.trim() === '') return '';
        dateStr = dateStr.trim();
        
        // Handle YYYYMMDD
        if (/^\d{8}$/.test(dateStr)) {
            return `${dateStr.substring(0, 4)}-${dateStr.substring(4, 6)}-${dateStr.substring(6, 8)}`;
        }
        
        // Try native Date parser
        const d = new Date(dateStr);
        if (!isNaN(d.getTime())) {
            const y = d.getFullYear();
            const m = String(d.getMonth() + 1).padStart(2, '0');
            const day = String(d.getDate()).padStart(2, '0');
            return `${y}-${m}-${day}`;
        }
        
        return dateStr;
    };

    const cleanBatch = (batch) => {
        if (!batch || batch.trim() === '') return 'UNKNOWN';
        batch = batch.trim().toUpperCase().replace(/-/g, '');
        if (batch.startsWith('BATCH')) {
            return 'B' + batch.substring(5);
        }
        return batch;
    };

    const stageMap = { '項目1': 'A', '項目2': 'B', '項目3': 'C' };
    const cleanStage = (stage) => stageMap[stage] || stage;

    const cleanTemp = (tempStr) => {
        if (!tempStr || tempStr.trim() === '') return '';
        const match = tempStr.match(/([-+]?\d*\.?\d+)/);
        if (!match) return '';
        let val = parseFloat(match[1]);
        if (tempStr.toUpperCase().includes('F')) {
            val = (val - 32) * 5 / 9;
        }
        return val.toFixed(1);
    };

    const timeMap = { '下午': '14:00-16:00', '晚': '18:00-20:00' };
    const cleanTime = (time) => timeMap[time] || time;

    const cleanedRows = rows.map(line => {
        const parts = line.split(',');
        const date = parseDate(parts[0]);
        const batch = cleanBatch(parts[1]);
        const stage = cleanStage(parts[2]);
        const temp = cleanTemp(parts[3]);
        const time = cleanTime(parts[4]);
        return `${date},${batch},${stage},${temp},${time}`;
    });

    // Remove duplicates
    const uniqueRows = [...new Set(cleanedRows)];
    
    // Sort
    uniqueRows.sort();

    const result = [header.join(','), ...uniqueRows].join('\n');
    fs.writeFileSync(outputPath, '\ufeff' + result, 'utf8'); // Add BOM for Excel compatibility
    console.log(`Cleaned data saved to ${outputPath}`);
}

cleanCsv('PRD-011.csv', 'PRD-011_cleaned.csv');

File: clean_data.py
import pandas as pd
import numpy as np
import re
from datetime import datetime

def clean_csv(file_path, output_path):
    # Read the CSV
    df = pd.read_csv(file_path)
    
    # 1. Standardize Date
    def parse_date(date_str):
        if pd.isna(date_str) or str(date_str).strip() == '':
            return None
        date_str = str(date_str).strip()
        formats = ['%Y%m%d', '%d %b %Y', '%Y/%m/%d', '%Y-%m-%d', '%Y/%n/%j']
        for fmt in formats:
            try:
                return datetime.strptime(date_str, fmt).strftime('%Y-%m-%d')
            except ValueError:
                continue
        # Try some manual fixes for single digit month/day
        try:
            return pd.to_datetime(date_str).strftime('%Y-%m-%d')
        except:
            return date_str

    df['日期'] = df['日期'].apply(parse_date)

    # 2. Standardize Batch Number (批號)
    def clean_batch(batch):
        if pd.isna(batch) or str(batch).strip() == '':
            return 'UNKNOWN'
        batch = str(batch).upper().replace('-', '')
        if batch.startswith('BATCH'):
            return 'B' + batch[5:]
        return batch

    df['批號'] = df['批號'].apply(clean_batch)

    # 3. Standardize Process Stage (製程段)
    # Mapping 項目1->A, 項目2->B, 項目3->C (Assuming this mapping based on general patterns)
    stage_map = {'項目1': 'A', '項目2': 'B', '項目3': 'C'}
    df['製程段'] = df['製程段'].replace(stage_map)

    # 4. Standardize Temperature (溫度) to Celsius
    def clean_temp(temp_str):
        if pd.isna(temp_str) or str(temp_str).strip() == '':
            return None
        temp_str = str(temp_str).upper()
        match = re.search(r'([-+]?\d*\.?\d+)', temp_str)
        if not match:
            return None
        val = float(match.group(1))
        if '°F' in temp_str or 'F' in temp_str:
            return round((val - 32) * 5 / 9, 1)
        return round(val, 1)

    df['溫度(°C)'] = df['溫度'].apply(clean_temp)
    df = df.drop(columns=['溫度'])

    # 5. Standardize Time Period (時段)
    # Mapping '下午' to '14:00-16:00', '晚' to '18:00-20:00' (Estimated)
    time_map = {'下午': '14:00-16:00', '晚': '18:00-20:00'}
    df['時段'] = df['時段'].replace(time_map)

    # 6. Remove Duplicates
    df = df.drop_duplicates()

    # 7. Final Polish
    # Sort by Date and Batch
    df = df.sort_values(by=['日期', '批號'])

    # Save to new CSV
    df.to_csv(output_path, index=False, encoding='utf-8-sig')
    print(f"Cleaned data saved to {output_path}")

if __name__ == "__main__":
    clean_csv('PRD-011.csv', 'PRD-011_cleaned.csv')
